Recursive icon lookup

fwSmit opened the pull request:
This PR implements recursive icon lookup.

This code is taken from mako's icon.c and only slightly adapted to use with dunst. Due to this change the icon_path should be changed to the root of the icon theme, instead of a subdirectory of it. The icon size may also be different from before, since the largest icon that still fits is used, instead of the first icon found. If this icon is too large you should change max_icon_size accordingly.

Note: Users have to change their dunstrc because of this PR. Instead of adding every required subdirectory of an icon theme, you only need to add the path to the icon theme. So that would be `/usr/share/icons/Adwaita` instead of `/usr/share/icons/Adwaita/action/16x16/actions`.
Fixes: #822 #157

garberw commented:
I wrote a test script.
It did not find any errors for any of the icons
in Adwaita.

Not sure exactly what you want it to do.
for icons bigger than 32x32
I think it bumps them back down to 32x32.

attached
can you rsvp so I know you got this please.

thanks.
[wg-dunst-test-icons.sh.zip](https://github.com/dunst-project/dunst/files/6475884/wg-dunst-test-icons.sh.zip)

fwSmit commented:
That's a pretty cool looking test! Only you're testing full icon paths here. If you send a full icon path it doesn't even need to be in your icon_path. I've changed your script to strip the path and it still works.
Here's my version: https://gist.github.com/fwSmit/cc3b3c5efb5f1f5c6647aaf7eb33ff5d

> Not sure exactly what you want it to do.
> for icons bigger than 32x32
> I think it bumps them back down to 32x32.

If you send in the full path, it should scale them down to be at most the size of `max_icon_size`.
If you're sending only the name, like in my version of the script, it should 
use the biggest icon that still fits `max_icon_size`. In my case I have the `max_icon_size` set to 128, so it returns icons of 96 pixels big, since the 256x256 icons are too big. I doesn't need to scale them, since my `min_icon_size` is set to 16, so this results in an icon of 96x96.

vincentbernat commented:
I am also pretty interested by this one. Do you care to rebase it?

vincentbernat commented:
Without testing:
* I would prefer to use the next biggest icon and scale it down. I am using `min_icon_size` and `max_icon_size` at the same value. I think you could filter out the sizes lesser than `min_icon_size`, then consider the max of the remaining ones inferior to `max_icon_size` and if there are none, choose the minimal one.
* Many themes also have a `scalable` size. After a quick look, it seems you don't consider it. I don't know if the spec says something about it. Maybe preferring it if there is no exact match?

vincentbernat commented:
FYI, I have rebased your PR on top of the current scaling branch. This gives the following commits:
* https://github.com/dunst-project/dunst/commit/445759362413f8693f0b16b9a452fc3f6eadbc79 (yours, with just what is needed to rebase)
* https://github.com/dunst-project/dunst/commit/dc7e456f034e106bd1ab5f1900b2c58f913ba9fc (mine to handle scaling, with the policy I proposed above)

I have left the debug code.

fwSmit commented:
> Without testing:
> 
>     * I would prefer to use the next biggest icon and scale it down. I am using `min_icon_size` and `max_icon_size` at the same value. I think you could filter out the sizes lesser than `min_icon_size`, then consider the max of the remaining ones inferior to `max_icon_size` and if there are none, choose the minimal one.

Yeah that would be a slight improvement.


>     * Many themes also have a `scalable` size. After a quick look, it seems you don't consider it. I don't know if the spec says something about it. Maybe preferring it if there is no exact match?

I believe this code is indeed skipping the scalable directory. It would also be an improvement to include that one if no suitable match is found.




> FYI, I have rebased your PR on top of the current scaling branch. This gives the following commits:
> 
>     * [4457593](https://github.com/dunst-project/dunst/commit/445759362413f8693f0b16b9a452fc3f6eadbc79) (yours, with just what is needed to rebase)
> 
>     * [dc7e456](https://github.com/dunst-project/dunst/commit/dc7e456f034e106bd1ab5f1900b2c58f913ba9fc) (mine to handle scaling, with the policy I proposed above)
> 
> 
> I have left the debug code.

Thank you. It seems like in your code you've addressed the first comment and added the scaling factor to the mix.
For the first one, I'm not sure if you have selected the simplest solution. It looks quite complicated.
For the second one, I don't think that it's right to multiply the icon size with the scale. Those icons at 2x scale are only meant for when you are actually viewing them on a HiDPI monitor. A 48x48@2x might be different from a 96x96 icon.

If you are curious, here's the icon spec: https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html.

I think I'll just use your rebased commit and leave the rest for others (or myself) to improve later.

fwSmit commented:
@tsipinakis This PR is ready now

fwSmit commented on src/icon.c:
```diff
@@ -208,65 +205,156 @@ GdkPixbuf *get_pixbuf_from_file(const char *filename, int scale)
+        /* if (output->scale > max_scale) { */
+        /* max_scale = output->scale; */
+        /* } */
+        /* } */
```
I implemented it, but ended up having to add another FIXME. I'm getting confused by the current way icons are requested, so I'll refactor that some time.

fwSmit commented on docs/dunst.5.pod:
```diff
@@ -443,13 +443,23 @@ gets the last say.
+you'll need to add them to the path manually.
+
+Please note that the path should now be the root of the icon theme, not a
+subdirectory of it. This has changed in a recent version and you should change
```
Thanks for the suggestion. I didn't know what version it was gonna be, but I guess it will be 1.7.

fwSmit commented:
I've improved the icon selection algorithm a bit, since it wasn't using a `32x32@2x` icon over `32x32` when icon_size > 32 and scale = 2.
What still could be improved is when icon size has to be `min <= size <= max` and no icon is found between size min and max, some bigger icon has to be chosen, instead of a smaller one. Currently `min_icon_size` is not considered in the selection algorithm.
I'll leave that for someone else to improve, though. So this PR is ready

OliverLew commented:
It will be even better if an icon theme is required, instead of a path. With hicolor and pixmaps as fallback, usually one icon theme is enough.
And look for the icon theme directory in `$XDG_DATE_HOME/icons` or `$XDG_DATA_DIRS/icons`. Those XDG directories fall back to `$HOME/.local/share` and `/usr/share:/usr/local/share`.

Zopieux commented:
+1 to using well-established lookup mechanisms like XDG rather than deploying yet another one, as long as sensible fallback locations are provided.

fwSmit commented:
> It will be even better if an icon theme name is required, instead of a path. With hicolor and pixmaps as fallback, usually one icon theme is enough.
> And look for the icon theme directory in `$XDG_DATE_HOME/icons` or `$XDG_DATA_DIRS/icons`. Those XDG directories fall back to `$HOME/.local/share` and `/usr/share:/usr/local/share`.

Agreed, I'm on it

fwSmit commented:
I'm noticing some performance problems with the recursive icon lookup, especially when using bigger icon theme. For example, searching the Papirus-dark icon theme takes ~0.15 seconds. This is on a fast ssd and even then it's noticable when using dunst for volume notifications. This will be an even bigger problem when allowing users to specify multiple icon themes (this can be avoided by allowing only 1 icon theme and making it ruleable).

OliverLew commented:
I don't know much about C, but if the glob search is the bottle neck here, you can try this:
- Parse the `index.theme` file. Within it are descriptions of all the subfolders, e.g. `64x64/places`
- Find the `Type` and `(Min/Max)Size`, they tell you what sizes this subfolder is for
- Find the subfolder(s) that match the desired icon size
- Check directly if `$theme_dir/$subfolder/icon.ext` exists.

This way there is no glob searching, and more close to the spec I think. I have a shell script implementing this method, which only uses ~0.07 seconds (I timed only the relevant code) to find one icon. I think it should be a lot faster in C.

fwSmit commented:
> I don't know much about C, but if the glob search is the bottle neck here, you can try this:
> 
>     * Parse the `index.theme` file. Within it are descriptions of all the subfolders, e.g. `64x64/places`
> 
>     * Find the `Type` and `(Min/Max)Size`, they tell you what sizes this subfolder is for
> 
>     * Find the subfolder(s) that match the desired icon size
> 
>     * Check directly if `$theme_dir/$subfolder/icon.ext` exists.
> 
> 
> This way there is no glob searching, and more close to the spec I think. I have a shell script implementing this method, which only uses ~0.07 seconds (I timed only the relevant code) to find one icon. I think it should be a lot faster in C.

That's a good idea. The [spec](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html) doesn't dictate the way you should look up icons, but gives an example similar to what I already have.
But let's say there is an icon that only exists with size (Min/Max): (16/32) and the minimum icon size configured in dunst is 64. Should dunst not use this icon? Or should dunst first look in all folders anyways if it can't find it?
I'd be interested to see that script. Would you be willing to share it?

OliverLew commented:
> The [spec](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html) doesn't dictate the way you should look up icons, but gives an example similar to what I already have.

Indeed.

> But let's say there is an icon that only exists with size (Min/Max): (16/32) and the minimum icon size configured in dunst is 64. Should dunst not use this icon? Or should dunst first look in all folders anyways if it can't find it?

Maybe yes. Normally the size limits are quite tolerative, The icon themes I have all have a MaxSize of either 256 or 512. The only exception is one subfolder in Adwaita icon theme, `scalable-up-to-32/status`, which has a MaxSize of, as expected, 32.

> I'd be interested to see that script. Would you be willing to share it?

https://github.com/OliverLew/xdg-xmenu. It's one script I wrote to find icons and generate menu for xmenu. The relevant code is in this part:
https://github.com/OliverLew/xdg-xmenu/blob/91cd9b8e007af5117efae4a531737456fa77bd03/xdg-xmenu#L47-L80

And all the way to line 109 for all the icon searching.

Just checked the code, and actually I didn't strictly follow the spec. That doesn't affect the speed estimation, tho.

fwSmit commented:
Thank you @OliverLew for your script and the rest for your input. I think I'll clean up the code a bit and merge this version for now. When I have some time I'll work on the performance improvements.

fwSmit commented:
Closing this one in favour of #947.
